[BUG] IndexOutOfBoundsException when joining on 2 integer columns with DPP

**Describe the bug**
When joining on 2 integer columns that a table is partitioned by using DPP, get an `IndexOutOfBoundsException` in `GpuSubqueryBroadcastExec.projectSerializedBatchToRows`. This happens both when AQE is enabled or disabled.

**Steps/Code to reproduce bug**

In `dpp_test.py`, you can add the following query to `_statements`, and run `test_dpp_reuse_broadcast_exchange` to produce the exception:

```
    SELECT fact.key, fact.skey, sum(fact.value)
    FROM {0} fact
    JOIN {1} dim
    ON fact.key = dim.key AND fact.skey = dim.skey
    WHERE dim.filter = {2}
    GROUP BY fact.key, fact.skey
```

Relevant backtrace:

```
E                   Caused by: java.lang.IndexOutOfBoundsException: 1
E                   	at scala.collection.LinearSeqOptimized.apply(LinearSeqOptimized.scala:67)
E                   	at scala.collection.LinearSeqOptimized.apply$(LinearSeqOptimized.scala:65)
E                   	at scala.collection.immutable.List.apply(List.scala:91)
E                   	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.$anonfun$projectSerializedBatchToRows$1(GpuSubqueryBroadcastExec.scala:232)
E                   	at scala.Option.map(Option.scala:230)
E                   	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.projectSerializedBatchToRows(GpuSubqueryBroadcastExec.scala:231)
E                   	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.$anonfun$relationFuture$2(GpuSubqueryBroadcastExec.scala:212)
E                   	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withExecutionId$1(SQLExecution.scala:145)
E                   	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:169)
E                   	at org.apache.spark.sql.execution.SQLExecution$.withExecutionId(SQLExecution.scala:143)
E                   	at org.apache.spark.sql.rapids.execution.GpuSubqueryBroadcastExec.$anonfun$relationFuture$1(GpuSubqueryBroadcastExec.scala:209)
E                   	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
E                   	at scala.util.Success.$anonfun$map$1(Try.scala:255)
E                   	at scala.util.Success.map(Try.scala:213)
E                   	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
E                   	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
E                   	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
E                   	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
E                   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
E                   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
```

**Additional context**

Upon further inspection of the plan, you will notice that Spark does the following optimization in the `HashedRelationBroadcastMode`, where it will join the 2 int key expressions together into a single bigint key expression, so this means that `GpuSubqueryBroadcast` would need to handle splitting that up.

```
                          :     :- GpuSubqueryBroadcast dynamicpruning#291, 0, [key#265, skey#266]
                           :     :  +- GpuBroadcastExchange HashedRelationBroadcastMode(List((shiftleft(cast(input[0, int, true] as bigint), 32) | (cast(input[1, int, true] as bigint) & 4294967295))),false), [id=#434]
```

See https://github.com/apache/spark/blob/29eca8c87f4e8c19c0380f7c30668fd88edee573/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala#L720 for join key rewrite

--- a/integration_tests/src/main/python/delta_zorder_test.py
+++ b/integration_tests/src/main/python/delta_zorder_test.py
@@ -60,6 +60,15 @@ _statements = [
     WHERE dim.filter = {2}
     GROUP BY fact.ex_key
     ''',
+    # join on 2 z-ordered columns
+    '''
+    SELECT fact.ex_key, fact.ex_skey, sum(fact.value)
+    FROM {0} fact
+    JOIN {1} dim
+    ON fact.ex_key = dim.ex_key AND fact.ex_skey = dim.ex_skey
+    WHERE dim.filter = {2}
+    GROUP BY fact.ex_key, fact.ex_skey
+    ''',
     # join on 1 partitioned and 1 z-ordered column
     '''
     SELECT fact.key, fact.ex_key, sum(fact.value)
@@ -115,9 +124,10 @@ def test_delta_dfp_reuse_broadcast_exchange(spark_tmp_table_factory, s_index, aq
         # Note that ex_key is a high-cardinality column, which makes it a good candidate for 
         # for Z-ordering, which then means it can then be used in Dynamic File Pruning in joins
         df = gen_df(spark, [
-            ('key', LongGen(nullable=False, min_val=0, max_val=9, special_cases=[])),
-            ('skey', LongGen(nullable=False, min_val=0, max_val=4, special_cases=[])),
-            ('ex_key', LongGen(nullable=False, min_val=0, max_val=10000, special_cases=[])),
+            ('key', IntegerGen(nullable=False, min_val=0, max_val=9, special_cases=[])),
+            ('skey', IntegerGen(nullable=False, min_val=0, max_val=4, special_cases=[])),
+            ('ex_key', IntegerGen(nullable=False, min_val=0, max_val=10000, special_cases=[])),
+            ('ex_skey', IntegerGen(nullable=False, min_val=0, max_val=1000, special_cases=[])),
             ('value', int_gen),
         ], 10000)
 
@@ -125,12 +135,13 @@ def test_delta_dfp_reuse_broadcast_exchange(spark_tmp_table_factory, s_index, aq
             .mode("overwrite") \
             .partitionBy("key", "skey") \
             .saveAsTable(fact_table)
-        spark.sql("OPTIMIZE {} ZORDER BY ex_key".format(fact_table)).show()
+        spark.sql("OPTIMIZE {} ZORDER BY (ex_key, ex_skey)".format(fact_table)).show()
 
         df = gen_df(spark, [
-            ('key', LongGen(nullable=False, min_val=0, max_val=9, special_cases=[])),
-            ('skey', LongGen(nullable=False, min_val=0, max_val=4, special_cases=[])),
-            ('ex_key', LongGen(nullable=False, min_val=0, max_val=10000, special_cases=[])),
+            ('key', IntegerGen(nullable=False, min_val=0, max_val=9, special_cases=[])),
+            ('skey', IntegerGen(nullable=False, min_val=0, max_val=4, special_cases=[])),
+            ('ex_key', IntegerGen(nullable=False, min_val=0, max_val=10000, special_cases=[])),
+            ('ex_skey', IntegerGen(nullable=False, min_val=0, max_val=1000, special_cases=[])),
             ('value', int_gen),
             ('filter', RepeatSeqGen(
                 IntegerGen(min_val=0, max_val=2000, special_cases=[]), length=2000 // 20))

--- a/integration_tests/src/main/python/dpp_test.py
+++ b/integration_tests/src/main/python/dpp_test.py
@@ -94,6 +94,14 @@ _statements = [
     GROUP BY f.key
     ''',
     '''
+    SELECT fact.key, fact.skey, sum(fact.value)
+    FROM {0} fact
+    JOIN {1} dim
+    ON fact.key = dim.key AND fact.skey = dim.skey
+    WHERE dim.filter = {2}
+    GROUP BY fact.key, fact.skey
+    ''',
+    '''
     SELECT fact.key, fact.skey, fact.ex_key, sum(fact.value)
     FROM {0} fact
     JOIN {1} dim

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala
@@ -228,9 +228,9 @@ object GpuHashJoin extends Arm {
   }
 
   /**
-    * Copied from Apache Spark, returns whether the keys can be rewritten as a packed long. If
-    * they can, we can assume that they are packed when we extract them out.
-    */
+   * Copied from Apache Spark, returns whether the keys can be rewritten as a packed long. If
+   * they can, we can assume that they are packed when we extract them out.
+   */
   def canRewriteAsLongType(keys: Seq[Expression]): Boolean = {
     keys.forall(_.dataType.isInstanceOf[IntegralType]) &&
       keys.map(_.dataType.defaultSize).sum <= 8

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala
@@ -227,11 +227,17 @@ object GpuHashJoin extends Arm {
     }
   }
 
+  // scalastyle:off line.size.limit
   /**
-   * Copied from Apache Spark, returns whether the keys can be rewritten as a packed long. If
+   * The function is copied from Spark 3.2:
+   *   https://github.com/apache/spark/blob/v3.2.2/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala#L709-L713
+   *
+   * Returns whether the keys can be rewritten as a packed long. If
    * they can, we can assume that they are packed when we extract them out.
    */
+  // scalastyle:on
   def canRewriteAsLongType(keys: Seq[Expression]): Boolean = {
+    // TODO: support BooleanType, DateType and TimestampType
     keys.forall(_.dataType.isInstanceOf[IntegralType]) &&
       keys.map(_.dataType.defaultSize).sum <= 8
   }

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala
@@ -226,6 +226,15 @@ object GpuHashJoin extends Arm {
       case _ => false
     }
   }
+
+  /**
+    * Copied from Apache Spark, returns whether the keys can be rewritten as a packed long. If
+    * they can, we can assume that they are packed when we extract them out.
+    */
+  def canRewriteAsLongType(keys: Seq[Expression]): Boolean = {
+    keys.forall(_.dataType.isInstanceOf[IntegralType]) &&
+      keys.map(_.dataType.defaultSize).sum <= 8
+  }
 }
 
 abstract class BaseHashJoinIterator(

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubqueryBroadcastExec.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubqueryBroadcastExec.scala
@@ -32,7 +32,7 @@ import org.apache.spark.sql.catalyst.plans.physical.IdentityBroadcastMode
 import org.apache.spark.sql.execution.{BaseSubqueryExec, SparkPlan, SQLExecution, SubqueryBroadcastExec}
 import org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec
 import org.apache.spark.sql.execution.exchange.BroadcastExchangeExec
-import org.apache.spark.sql.execution.joins.HashedRelationBroadcastMode
+import org.apache.spark.sql.execution.joins.{HashedRelationBroadcastMode, HashJoin}
 import org.apache.spark.util.ThreadUtils
 
 
@@ -229,14 +229,28 @@ case class GpuSubqueryBroadcastExec(
     // are being extracted. The CPU already has the key projections applied in the broadcast
     // data and thus does not have similar logic here.
     val broadcastModeProject = modeKeys.map { keyExprs =>
-      val keyExpr = keyExprs(index)
+      val keyExpr = if (GpuHashJoin.canRewriteAsLongType(buildKeys)) {
+        // in this case, there is only 1 key expression since it's a packed version that encompasses
+        // multiple integral values into a single long using bit logic. In CPU Spark, the broadcast
+        // would create a LongHashedRelation instead of a standard HashedRelation.
+        keyExprs(0)
+      } else {
+        keyExprs(index)
+      }
       UnsafeProjection.create(keyExpr)
     }
 
     // Use the single output of the broadcast mode projection if it exists
     val rowProjectIndex = if (broadcastModeProject.isDefined) 0 else index
-    val rowProject = UnsafeProjection.create(
-      BoundReference(rowProjectIndex, buildKeys(index).dataType, buildKeys(index).nullable))
+    val rowExpr = if (GpuHashJoin.canRewriteAsLongType(buildKeys)) {
+      // Since this is the expected output for a LongHashedRelation, we can extract the key from the
+      // long packed key using bit logic, using this method available in HashJoin to give us the 
+      // correct key expression. 
+      HashJoin.extractKeyExprAt(buildKeys, index)
+    } else {
+      BoundReference(rowProjectIndex, buildKeys(index).dataType, buildKeys(index).nullable)
+    }
+    val rowProject = UnsafeProjection.create(rowExpr)
 
     // Deserializes the batch on the host. Then, transforms it to rows and performs row-wise
     // projection. We should NOT run any device operation on the driver node.

--- a/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubqueryBroadcastExec.scala
+++ b/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubqueryBroadcastExec.scala
@@ -233,7 +233,7 @@ case class GpuSubqueryBroadcastExec(
         // in this case, there is only 1 key expression since it's a packed version that encompasses
         // multiple integral values into a single long using bit logic. In CPU Spark, the broadcast
         // would create a LongHashedRelation instead of a standard HashedRelation.
-        keyExprs(0)
+        keyExprs.head
       } else {
         keyExprs(index)
       }

